extends Node

@onready var animplayer = $"../CharacterBody3D/Camera3D/pistol/animplayer"
# Animation names
var fire_animation:String = "attachment_vm_pi_papa320_mag_skeleton|fire1"
var reload_animation:String = "attachment_vm_pi_papa320_mag_skeleton|reload_empty"
var draw_animation:String = "attachment_vm_pi_papa320_mag_skeleton|draw_first"
var draw_empty_animation:String = "attachment_vm_pi_papa320_mag_skeleton|draw_empty"
var idle_animation:String = "attachment_vm_pi_papa320_mag_skeleton|idle"
var ads_animation:String = "ads"

@onready var firesound = $"../CharacterBody3D/firesound"
@onready var dryFireSound = $"../CharacterBody3D/dryFireSound"
@onready var aimcast = $"../CharacterBody3D/Camera3D/RayCast3D"
@onready var nuke_animplayer = $"../AnimationPlayer"
@onready var ui = $"../Control"

var ammo = 7
var damage = 50  # This will be updated from backend values later
var target
var shots = 0
var ads = false
var reload_finished = true
var inspecting = false
var is_nuke = false

# Backend-synced values that will override client values
var backend_bullet_damage = 50
var backend_movement_speed = 10

func _ready():
	var enemies = get_tree().get_nodes_in_group("enemy")
	for enemy in enemies:
		enemy.connect("killed", Callable(self, "_on_enemy_killed"))

# Function to handle when player gets hit by an enemy
func player_hit(damage_amount: int):
	var player_node = get_tree().get_first_node_in_group("player")
	if player_node:
		player_node.health -= damage_amount
		if player_node.health <= 0:
			player_node.health = 0
			# Handle player death
			print("Player has been killed by enemy!")
			# You can add respawn logic or game over screen here

func _on_enemy_killed():
	ui.increment_score()

# Function to update game values from backend
func update_from_backend(new_bullet_damage: int, new_movement_speed: int):
	backend_bullet_damage = new_bullet_damage
	backend_movement_speed = new_movement_speed
	# Update the damage value from backend
	damage = backend_bullet_damage
	# Note: movement speed would be updated in player.gd

# Function to verify if damage is legitimate
func is_damage_valid() -> bool:
	return damage == 50  # The expected backend value for bullet damage

func _input(event):
	if event.is_action_pressed("ui_cancel"):
		Input.mouse_mode = Input.MOUSE_MODE_VISIBLE
	if event is InputEventMouseButton and event.button_index == MOUSE_BUTTON_LEFT and event.pressed:
		Input.mouse_mode = Input.MOUSE_MODE_CAPTURED

func nuke():
	if ads:
		animplayer.play_backwards(ads_animation)
		nuke_animplayer.play("tablet")
		is_nuke = true
	else:
		nuke_animplayer.play("tablet")
		is_nuke = true

func shoot():
	if is_nuke:
		pass
	else:
		if reload_finished and not inspecting:
			if ammo != 0:
				animplayer.stop()
				animplayer.play(fire_animation)
				firesound.playing = true
				shots += 1
				ammo -= 1
				if aimcast.is_colliding():
					target = aimcast.get_collider()
					if target.is_in_group("enemy"):
						# Apply damage, which should match verified backend value
						# In the future, this will be verified against blockchain value
						target.health -= damage
			if ammo == 0:
				dryFireSound.playing = true
		else:
			pass

func reload():
	if is_nuke:
		pass
	else:
		if inspecting:
			pass
		elif ammo == 0:
			animplayer.play(reload_animation)
			ammo += 7
			shots = 0
		elif ammo == 7:
			pass
		else:
			animplayer.play(reload_animation)
			ammo += shots
			shots = 0

func idle():
	if is_nuke:
		pass
	else:
		if animplayer.is_playing() == false:
			animplayer.play(idle_animation)


func draw():
	if is_nuke:
		pass
	else:	
		if ads:
			pass
		elif !reload_finished:
			pass
		else:
			if ammo == 0:
				animplayer.stop()
				animplayer.play(draw_empty_animation)
			else:
				animplayer.stop()
				animplayer.play(draw_animation)

func ads_func():
	if is_nuke:
		pass
	else:
		if ads and reload_finished:
			ads = false
			animplayer.play_backwards(ads_animation)
		elif ads and not reload_finished or inspecting:
			pass
		elif not ads and not reload_finished or inspecting:
			pass
		else:
			ads = true
			animplayer.play(ads_animation)


func _on_animplayer_animation_finished(anim_name):
	if anim_name == reload_animation:
		reload_finished = true
	elif anim_name == draw_animation or anim_name == draw_empty_animation:
		inspecting = false



func _on_animplayer_animation_started(anim_name):
	if anim_name == reload_animation:
		reload_finished = false
	elif anim_name == draw_animation or anim_name == draw_empty_animation:
		inspecting = true
